### ISSUE TITLE
Add a property to ignore guard clauses in ReturnCountCheck

### ISSUE THREAD

**Issue by stoyanK7** [checkstyle/checkstyle#12049]
## The check
https://checkstyle.sourceforge.io/config_coding.html#ReturnCount

## Definition of guard clause
[From Wikipedia article](https://en.wikipedia.org/wiki/Guard_(computer_science)):

>Regardless of which programming language is used, a guard clause, guard code, or guard statement, is a check of integrity [preconditions](https://en.wikipedia.org/wiki/Precondition) used to avoid errors during execution. A typical example is checking that a reference about to be processed is not null, which avoids null-pointer failures. Other uses include using a boolean field for [idempotence](https://en.wikipedia.org/wiki/Idempotence) (so subsequent calls are nops), as in the [dispose pattern](https://en.wikipedia.org/wiki/Dispose_pattern). The guard provides an [early exit](https://en.wikipedia.org/wiki/Early_exit) from a [subroutine](https://en.wikipedia.org/wiki/Subroutine), and is a commonly used deviation from [structured programming](https://en.wikipedia.org/wiki/Structured_programming), removing one level of nesting and resulting in flatter code..

## Problem description
Add a property `ignoreGuardClauses`(for example) that would ignore any `if` statement with just a `return` inside it. If the `if` statement has more than just a `return` statment inside it, then it would be considered part of the method logic(not a guard clause) and the `return` should be counted.

Currently this is not possible and nested `if` statments might be confusing and hard to follow.
 
#### Example
I would like the following..
```java
	@Override
	public boolean accept(AuditEvent event) {
		boolean accepted = true;
		if (event.getViolation() != null) {
			final FileText fileText = getFileText(event.getFileName());
			if (fileText != null) {
				final List<Suppression> suppressions = getSuppressions(fileText);
				accepted = getNearestSuppression(suppressions, event) == null;
			}
		}
		return accepted;
	}


```
to be turned into..
```java
	@Override
	public boolean accept(AuditEvent event) {
		if (event.getViolation() == null) return true;

		final FileText fileText = getFileText(event.getFileName());
		if (fileText == null) return true;

		final List<Suppression> suppressions = getSuppressions(fileText);
		return getNearestSuppression(suppressions, event) == null;
	}


```
..without any violations. I believe this looks much neater.

---
##### Current output
```shell
stoyan@ZenBook:~/Desktop/checkstyle$ cat config.xml 
```
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN" "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="ReturnCount">
      <property name="max" value="1"/>
    </module>
  </module>
</module>
```

```shell
stoyan@ZenBook:~/Desktop/checkstyle$ cat Test.java
```
```java
public class Test {

    public boolean method(AuditEvent event) {
        final FileText fileText = getFileText(event.getFileName());
        if (fileText == null) { 
            return true; // return count = 1
        }

        if (fileText == null)
             return true; // return count = 2

        if (fileText == null) { return true; } // return count = 3

        if (fileText == null) return true; // return count = 4

        return true; // return count = 5
    }

    public int search(String[] arr, String strToSearchFor) {
        for(int index = 0; index < arr.length; index++) {
            if (arr[index].equals(strToSearchFor)) {
                return index; // return count = 1
            } else if (isBackwards(arr[index], strToSearchFor)) {
                return -index; // return count = 2
            }
        }
        return -999; // return count = 3
    }

    public int someBadExample(int number) {
        int someVar = 30;

        if (number > 5000) {
            doSomeThingWithNumber(number);
            return 1; // return count = 1
        } else if (someVar - number > 30) {
            this.checkLength("blabla");
            someVar--;
            return -1; // return count = 2;
        }

        return 100; // return count = 3;
    }

    public String checkLength(String str) {
        if (str.length > 50) {
            return "String is too long"; // return count = 1
        } else if (str.length > 20) {
            return "String is medium long"; // return count = 2
        } else {
            return "String is 20 or less characters long"; // return count = 3
        }
    }
}
```

```shell
stoyan@ZenBook:~/Desktop/checkstyle$ java -jar checkstyle.jar -c config.xml Test.java 
```
```shell
Starting audit...
[ERROR] /home/stoyan/Desktop/checkstyle/Test.java:3:9: Return count is 5 (max allowed for non-void methods/lambdas is 1). [ReturnCount]
[ERROR] /home/stoyan/Desktop/checkstyle/Test.java:19:5: Return count is 3 (max allowed for non-void methods/lambdas is 1). [ReturnCount]
[ERROR] /home/stoyan/Desktop/checkstyle/Test.java:30:5: Return count is 3 (max allowed for non-void methods/lambdas is 1). [ReturnCount]
[ERROR] /home/stoyan/Desktop/checkstyle/Test.java:45:5: Return count is 3 (max allowed for non-void methods/lambdas is 1). [ReturnCount]


```
##### Desired output
```shell
stoyan@ZenBook:~/Desktop/checkstyle$ cat config.xml 
```
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN" "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="ReturnCount">
      <property name="max" value="1"/>
      <property name="ignoreGuardClauses" value="true"/>
    </module>
  </module>
</module>


```

```shell
stoyan@ZenBook:~/Desktop/checkstyle$ cat Test.java 
```
```java
public class Test {

    public boolean method(AuditEvent event) {
        final FileText fileText = getFileText(event.getFileName());
        if (fileText == null) { 
            return true; // return count = 0
        }

        if (fileText == null)
             return true; // return count = 0

        if (fileText == null) { return true; } // return count = 0

        if (fileText == null) return true; // return count = 0

        return true; // return count = 1
    }

    public int search(String[] arr, String strToSearchFor) {
        for(int index = 0; index < arr.length; index++) {
            if (arr[index].equals(strToSearchFor)) {
                return index; // return count = 0
            } else if (isBackwards(arr[index], strToSearchFor)) {
                return -index; // return count = 0
            }
        }
        return -999; // return count = 1
    }

    public int someBadExample(int number) {
        int someVar = 30;

        if (number > 5000) {
            doSomeThingWithNumber(number);
            return 1; // return count = 1
        } else if (someVar - number > 30) {
            this.checkLength("blabla");
            someVar--;
            return -1; // return count = 2;
        }

        return 100; // return count = 3;
    }

    public String checkLength(String str) {
        if (str.length > 50) {
            return "String is too long"; // return count = 0
        } else if (str.length > 20) {
            return "String is medium long"; // return count = 0
        } else {
            return "String is 20 or less characters long"; // return count = 1
        }
    }
}

```

```shell
stoyan@ZenBook:~/Desktop/checkstyle$ java -jar checkstyle.jar -c config.xml Test.java 
```
```shell
Starting audit...
[ERROR] /home/stoyan/Desktop/checkstyle/Test.java:30:5: Return count is 3 (max allowed for non-void methods/lambdas is 1). [ReturnCount]


```

**Comment by strkkk** [comment]
@stoyanK7 it is not so clear how to recognize `guardClause`. It can lead to false positives.

For example, in your code snippet, second if check
`if (fileText == null) return true;`
is questionable.

Is there some clear definition of `guard clause`?

**Comment by stoyanK7** [comment]
@strkkk Taken from [wikipedia definition of guard(programming)](https://en.wikipedia.org/wiki/Dispose_pattern)
>The guard provides an [early exit](https://en.wikipedia.org/wiki/Early_exit) from a [subroutine](https://en.wikipedia.org/wiki/Subroutine), and is a commonly used deviation from [structured programming](https://en.wikipedia.org/wiki/Structured_programming), removing one level of nesting and resulting in flatter code

**Comment by nick-mancuso** [comment]
@stoyanK7 please add to your example in issue description (with expected behavior):
1. Similar `if` statements WITH `{ }` 
```java
		if (fileText == null) { return true; }

```
2. `if` statements that serve as "guard clauses" for exceptions
3. Multiline "guard clauses" both with and without `{}`
```java
		if (fileText == null) { 
                   return true; 
                }

		if (fileText == null)  
                    return true; 


```

> @strkkk Taken from [wikipedia definition of guard(programming)](https://en.wikipedia.org/wiki/Dispose_pattern)...

Please add explanation of guard clause to the issue description, as well.

**Comment by romani** [comment]
Please look at http://sevntu-checkstyle.github.io/sevntu.checkstyle/apidocs/com/github/sevntu/checkstyle/checks/coding/ReturnCountExtendedCheck.html

**Comment by stoyanK7** [comment]
@romani 
I see. If i understand correctly, property `minIgnoreReturnDepth`(`Sets the minimum "return" statement depth with that will be skipped by check.`) and `topLinesToIgnoreCount` can accomplish this. Thanks!
